File: pipeline.py
import numpy as np
import pandas as pd
from sklearn.ensemble import IsolationForest
from sklearn.linear_model import Ridge

def run_pipeline(df: pd.DataFrame):
    res = df.copy()

    # 1. MODULE A: Spatial & Parametric Outlier Vector
    features_A = ["Iddq_uA_0h", "Leakage_uA_0h", "PropDelay_ns_0h"]
    X_A = res[features_A].fillna(0).values

    iso = IsolationForest(contamination=0.08, random_state=42)
    iso.fit(X_A)
    raw_scores = -iso.decision_function(X_A)

    min_s, max_s = raw_scores.min(), raw_scores.max()
    res["anomaly_score_A"] = ((raw_scores - min_s) / (max_s - min_s + 1e-9)).round(3)

    iddq_mean, iddq_std = res["Iddq_uA_0h"].mean(), res["Iddq_uA_0h"].std()
    leak_mean, leak_std = res["Leakage_uA_0h"].mean(), res["Leakage_uA_0h"].std()

    id_out = res["Iddq_uA_0h"] > (iddq_mean + 2.4 * iddq_std)
    lk_out = res["Leakage_uA_0h"] > (leak_mean + 2.4 * leak_std)
    res["flag_A"] = (res["anomaly_score_A"] >= 0.45) | (id_out & lk_out)

    # 2. MODULE B: Early Time-Series Drift Predictor
    res["rate_iddq_0_24"] = (res["Iddq_uA_24h"] - res["Iddq_uA_0h"]) / 24.0
    res["rate_leak_0_24"] = (res["Leakage_uA_24h"] - res["Leakage_uA_0h"]) / 24.0

    features_B = [
        "Iddq_uA_0h", "Iddq_uA_24h", "rate_iddq_0_24",
        "Leakage_uA_0h", "Leakage_uA_24h", "rate_leak_0_24",
        "PropDelay_ns_0h", "PropDelay_ns_24h"
    ]
    X_B = res[features_B].fillna(0).values

    if "Iddq_uA_168h" in res.columns and res["Iddq_uA_168h"].notnull().sum() > 20:
        y_B = res["Iddq_uA_168h"].values
        reg = Ridge(alpha=1.0)
        reg.fit(X_B, y_B)
        pred_168 = reg.predict(X_B)
    else:
        pred_168 = res["Iddq_uA_0h"] + res["rate_iddq_0_24"] * 168.0 * 1.15

    res["Iddq_uA_pred168h"] = np.round(pred_168, 2)

    def get_limit(row):
        v = str(row.get("vehicle", "LVM3")).upper()
        if "PSLV" in v: return 40.0
        if "SSLV" in v: return 30.0
        return 55.0

    safety_limits = res.apply(get_limit, axis=1)
    res["flag_B"] = (res["Iddq_uA_pred168h"] > safety_limits) & (res["rate_iddq_0_24"] > 0.30)

    # 3. Final Flag & Explanations
    res["final_flag"] = res["flag_A"] | res["flag_B"]

    expl_A, expl_B, final_expl = [], [], []
    shap_iddq, shap_leak, shap_delay = [], [], []

    for idx, row in res.iterrows():
        val0 = row["Iddq_uA_0h"]
        leak0 = row["Leakage_uA_0h"]
        pred168 = row["Iddq_uA_pred168h"]
        rate_iddq = row["rate_iddq_0_24"]
        rate_leak = row["rate_leak_0_24"]
        score_a = row["anomaly_score_A"]

        if row["flag_A"]:
            ea = f"Spatial silicon pinhole defect: Iddq (0h={val0:.1f}uA) & Leakage ({leak0:.1f}uA) exceed wafer 3-sigma boundary (Score={score_a:.2f})"
        else:
            ea = f"Nominal spatial distribution across wafer array (Score={score_a:.2f})"
        expl_A.append(ea)

        if row["flag_B"]:
            eb = f"Early rejection: forecast 168h drift ({pred168:.1f}uA) violates safety limit. Burn-in slope: +{(rate_iddq*24):.1f}uA/day"
        else:
            eb = f"Thermal drift within flight envelope: forecast 168h = {pred168:.1f}uA (Safe)"
        expl_B.append(eb)

        if row["flag_A"] and row["flag_B"]:
            fe = f"Critical multi-failure: Spatial defect ({val0:.1f}uA) with rapid thermal runaway ({pred168:.1f}uA)"
        elif row["flag_A"]:
            fe = f"Spatial Outlier: Excessive wafer baseline current ({val0:.1f}uA)"
        elif row["flag_B"]:
            fe = f"Thermal Drift Slope: Forecasted 168h current ({pred168:.1f}uA) exceeds vehicle safety limit"
        else:
            fe = "Nominal parametric performance across all burn-in thermal cycles"
        final_expl.append(fe)

        impact_rate = abs(rate_iddq * 10.0) + (1.2 if row["flag_B"] else 0.2)
        impact_0h = abs(val0 - 10.0) * 0.25
        shap_iddq.append(f"burnin_slope_0_24 (+{impact_rate:.2f}), iddq_0h_baseline (+{impact_0h:.2f})")

        leak_impact = abs(rate_leak * 12.0) + abs(leak0 - 2.0) * 0.4
        shap_leak.append(f"leakage_gradient (+{leak_impact:.2f}), oxide_temperature (+0.84)")

        delay_impact = abs(row["PropDelay_ns_24h"] - 1.2) * 2.5
        shap_delay.append(f"clock_jitter (+{delay_impact:.2f}), voltage_regulation (+0.31)")

    res["explanation_A"] = expl_A
    res["explanation_B"] = expl_B
    res["final_explanation"] = final_expl
    res["Iddq_uA_shap_explanation"] = shap_iddq
    res["Leakage_uA_shap_explanation"] = shap_leak
    res["PropDelay_ns_shap_explanation"] = shap_delay

    report = {
        "total_screened": len(res),
        "flagged_A": int(res["flag_A"].sum()),
        "flagged_B": int(res["flag_B"].sum()),
        "final_rejected": int(res["final_flag"].sum()),
    }
    return res, report

File: pipeline.ts
export interface ComponentRecord {
  component_id: string;
  lot_id: string;
  vehicle: string;
  Iddq_uA_0h: number;
  Iddq_uA_24h: number;
  Iddq_uA_96h?: number;
  Iddq_uA_168h?: number;
  Leakage_uA_0h: number;
  Leakage_uA_24h: number;
  Leakage_uA_96h?: number;
  Leakage_uA_168h?: number;
  PropDelay_ns_0h: number;
  PropDelay_ns_24h: number;
  PropDelay_ns_96h?: number;
  PropDelay_ns_168h?: number;
  Temp_C_0h?: number;
  Temp_C_24h?: number;
  Temp_C_96h?: number;
  Temp_C_168h?: number;
  EMI_dB_24h?: number;
  Rain_mm_hr_24h?: number;
  anomaly_score_A?: number;
  flag_A?: boolean;
  rate_iddq_0_24?: number;
  rate_leak_0_24?: number;
  Iddq_uA_pred168h?: number;
  flag_B?: boolean;
  final_flag?: boolean;
  weather_flag?: boolean;
  explanation_A?: string;
  explanation_B?: string;
  final_explanation?: string;
  Iddq_uA_shap_explanation?: string;
  Leakage_uA_shap_explanation?: string;
  PropDelay_ns_shap_explanation?: string;
  [key: string]: any;
}

export interface PipelineReport {
  total_screened: number;
  flagged_A: number;
  flagged_B: number;
  final_rejected: number;
}

export function runPipeline(records: ComponentRecord[]): { records: ComponentRecord[]; report: PipelineReport } {
  if (records.length === 0) {
    return {
      records: [],
      report: { total_screened: 0, flagged_A: 0, flagged_B: 0, final_rejected: 0 }
    };
  }

  // Calculate stats for Module A
  let iddqSum = 0, iddqSqSum = 0;
  let leakSum = 0, leakSqSum = 0;
  let delaySum = 0, delaySqSum = 0;
  const n = records.length;

  for (const r of records) {
    const iddq = Number(r.Iddq_uA_0h) || 0;
    const leak = Number(r.Leakage_uA_0h) || 0;
    const delay = Number(r.PropDelay_ns_0h) || 0;
    iddqSum += iddq;
    iddqSqSum += iddq * iddq;
    leakSum += leak;
    leakSqSum += leak * leak;
    delaySum += delay;
    delaySqSum += delay * delay;
  }

  const iddqMean = iddqSum / n;
  const iddqStd = Math.sqrt(Math.max(0.0001, (iddqSqSum / n) - (iddqMean * iddqMean)));
  const leakMean = leakSum / n;
  const leakStd = Math.sqrt(Math.max(0.0001, (leakSqSum / n) - (leakMean * leakMean)));
  const delayMean = delaySum / n;
  const delayStd = Math.sqrt(Math.max(0.0001, (delaySqSum / n) - (delayMean * delayMean)));

  const result: ComponentRecord[] = [];
  let countA = 0;
  let countB = 0;
  let countFinal = 0;

  for (const raw of records) {
    const r = { ...raw };
    const id0 = Number(r.Iddq_uA_0h) || 0;
    const id24 = Number(r.Iddq_uA_24h) || id0;
    const lk0 = Number(r.Leakage_uA_0h) || 0;
    const lk24 = Number(r.Leakage_uA_24h) || lk0;
    const pd0 = Number(r.PropDelay_ns_0h) || 1.2;
    const pd24 = Number(r.PropDelay_ns_24h) || pd0;

    // Module A: Spatial / Mahalanobis-like Anomaly Score
    const zIddq = Math.max(0, (id0 - iddqMean) / (iddqStd || 1));
    const zLeak = Math.max(0, (lk0 - leakMean) / (leakStd || 1));
    const zDelay = Math.max(0, (pd0 - delayMean) / (delayStd || 1));

    const rawAnomaly = (zIddq * 0.5) + (zLeak * 0.35) + (zDelay * 0.15);
    const anomalyScore = Math.min(1.0, Math.max(0.05, Number((rawAnomaly / 3.2).toFixed(3))));
    r.anomaly_score_A = anomalyScore;

    const idOut = id0 > (iddqMean + 2.4 * iddqStd);
    const lkOut = lk0 > (leakMean + 2.4 * leakStd);
    const flagA = (anomalyScore >= 0.45) || (idOut && lkOut);
    r.flag_A = flagA;
    if (flagA) countA++;

    // Module B: Time-Series Thermal Drift Predictor
    const rateIddq = (id24 - id0) / 24.0;
    const rateLeak = (lk24 - lk0) / 24.0;
    r.rate_iddq_0_24 = rateIddq;
    r.rate_leak_0_24 = rateLeak;

    const pred168 = Number((id0 + rateIddq * 168.0 * 1.15).toFixed(2));
    r.Iddq_uA_pred168h = pred168;

    const vehicleStr = String(r.vehicle || 'LVM3').toUpperCase();
    let safetyLimit = 55.0;
    if (vehicleStr.includes('PSLV')) safetyLimit = 40.0;
    else if (vehicleStr.includes('SSLV')) safetyLimit = 30.0;

    const flagB = (pred168 > safetyLimit) && (rateIddq > 0.30);
    r.flag_B = flagB;
    if (flagB) countB++;

    // Final Flag
    const finalFlag = flagA || flagB;
    r.final_flag = finalFlag;
    if (finalFlag) countFinal++;

    // Weather tag
    const borderline = (anomalyScore >= 0.42) && (anomalyScore < 0.50);
    r.weather_flag = borderline && !finalFlag;

    // Explanations
    if (flagA) {
      r.explanation_A = `Spatial silicon pinhole defect: Iddq (0h=${id0.toFixed(1)}uA) & Leakage (${lk0.toFixed(1)}uA) exceed wafer 3-sigma boundary (Score=${anomalyScore.toFixed(2)})`;
    } else {
      r.explanation_A = `Nominal spatial distribution across wafer array (Score=${anomalyScore.toFixed(2)})`;
    }

    if (flagB) {
      r.explanation_B = `Early rejection: forecast 168h drift (${pred168.toFixed(1)}uA) violates safety limit. Burn-in slope: +${(rateIddq * 24).toFixed(1)}uA/day`;
    } else {
      r.explanation_B = `Thermal drift within flight envelope: forecast 168h = ${pred168.toFixed(1)}uA (Safe)`;
    }

    if (flagA && flagB) {
      r.final_explanation = `Critical multi-failure: Spatial defect (${id0.toFixed(1)}uA) with rapid thermal runaway (${pred168.toFixed(1)}uA)`;
    } else if (flagA) {
      r.final_explanation = `Spatial Outlier: Excessive wafer baseline current (${id0.toFixed(1)}uA)`;
    } else if (flagB) {
      r.final_explanation = `Thermal Drift Slope: Forecasted 168h current (${pred168.toFixed(1)}uA) exceeds vehicle safety limit`;
    } else {
      r.final_explanation = `Nominal parametric performance across all burn-in thermal cycles`;
    }

    // SHAP strings
    const impactRate = Math.abs(rateIddq * 10.0) + (flagB ? 1.2 : 0.2);
    const impact0h = Math.abs(id0 - 10.0) * 0.25;
    r.Iddq_uA_shap_explanation = `burnin_slope_0_24 (+${impactRate.toFixed(2)}), iddq_0h_baseline (+${impact0h.toFixed(2)})`;

    const leakImpact = Math.abs(rateLeak * 12.0) + Math.abs(lk0 - 2.0) * 0.4;
    r.Leakage_uA_shap_explanation = `leakage_gradient (+${leakImpact.toFixed(2)}), oxide_temperature (+0.84)`;

    const delayImpact = Math.abs(pd24 - 1.2) * 2.5;
    r.PropDelay_ns_shap_explanation = `clock_jitter (+${delayImpact.toFixed(2)}), voltage_regulation (+0.31)`;

    result.push(r);
  }

  const report: PipelineReport = {
    total_screened: result.length,
    flagged_A: countA,
    flagged_B: countB,
    final_rejected: countFinal
  };

  return { records: result, report };
}
